test/slot_test.exs

```diff
@@ -378,6 +426,118 @@ defmodule Surface.SlotTest do
            """
   end
 
+  test "slotable component with render defined" do
+    assigns = %{items: [%{id: 1, name: "First"}, %{id: 2, name: "Second"}]}
+
+    html =
+      render_surface do
+        ~H"""
+        <Grid items={{ user <- @items }}>
+          <ColumnWithRender title="column title">
+            <b>Id: {{ user.id }}</b>
+          </ColumnWithRender>
+        </Grid>
+        """
+      end
+
+    assert html =~ """
+           <table>
+             <tr>
+               <th>
+                 column title
+               </th>
+             </tr>
+             <tr>
+               <td>
+                 <span class="fancy-column">
+               <b>Id: 1</b>
+           </span>
+               </td>
+             </tr><tr>
+               <td>
+                 <span class="fancy-column">
+               <b>Id: 2</b>
+           </span>
+               </td>
+             </tr>
+           </table>
+           """
+  end
+
+  test "slotable component with slot props" do
+    assigns = %{items: [%{id: 1, name: "First"}, %{id: 2, name: "Second"}]}
+
+    html =
+      render_surface do
+        ~H"""
+        <Grid items={{ user <- @items }}>
+          <ColumnWithRenderAndSlotProps title="column title" :let={{ info: info }}>
+            <b>{{ info }}: {{ user.id }}</b>
```

Comment:
as is, this exposes the `Grid` prop named info; however, `ColumnWithRenderAndSlotProps` also has a prop named info.

@msaraiva @miguel-s @Malian any thoughts on what the expected behaviour would be in this situation?

Reply:
@lnr0626 I believe the Grid's `info` prop is not passed automatically to the slotables so there should be no conflict. The info var retrieved by `:let` should always be the one explicitly passed by `:props` inside `ColumnWithRenderAndSlotProps `. Right? If the user needs the Grid's `info` if has to retrieve it explicitly creating another variable with a different name, like:

```
<Grid items={{ user <- @items }} :let={{ info: grid_info }}>
```

As far as I can remember, only `user` will be passed automatically so if it happens to have the same name, we should have some kind of warning related to variable shadowing or something similar. Isn't the Elixir compiler already doing that for us?

Reply:
sort of - the issue here is the info prop on the grid component is on the cols slot, not the default slot; and the info slot on the ColumnWithRenderAndSlotProps is for its default slot. Because we're enabling renderable slotables, we now have two different slots that can each have their own props (in this case, `Grid.cols` and `ColumnWithRenderAndSlotProps.default`) where it's unclear which slot's props the `:let` would refer to. With a non-renderable component, we could do something like:

```
<Grid items={{ user <- @items }}>
  <template name="cols" :let={{ info: grid_info }}>
    <ColumnWithRenderAndSlotProps title="column title" :let={{ info: col_info }}>
      ...
    </ColumnWithRenderAndSlotProps>
  </template>
</Grid>
```

however, because this makes the slotable components combine the `template` and `<Component .../>` bits, we currently don't have a way to `:let` one of the two. With how it's currently setup, `:let` on a slotable component will expose props from the parent component, not the slotable component's default slot.

the relevant bits of those components:
```
  defmodule Grid do
    use Surface.Component

    prop items, :list, required: true

    slot cols, props: [:info, item: ^items]
    ...
  end

  defmodule ColumnWithRenderAndSlotProps do
    use Surface.Component, slot: "cols"

    prop title, :string, required: true

    slot default, props: [:info]
    ...
  end
```

this is really hard to describe clearly :-/

Reply:
playing around with svelte, it looks like their behavior is close to what's happening currently (i.e. let exposes the parent's props for the parents named slot, and the inner components default slot props aren't accessible):

Grid.svelte:
```
<div style="border: 1px solid black; padding: 4px">
	<slot name="content" info="info from grid" />
</div>
```

Column.svelte:
```
<div style="border: 1px solid blue; padding: 4px; margin: 4px">
	<slot info="info from column" />
</div>
```

App.svelte:
```
<script>
	import Grid from './Grid.svelte';
	import Column from './Column.svelte';
</script>

<Grid>
	<Column slot="content" let:info>
		{ info }
	</Column>
</Grid>

<Column let:info>
	{ info }
</Column>
```

results in:
<img width="522" alt="image" src="https://user-images.githubusercontent.com/704837/110692287-1197b100-819b-11eb-80ef-c8e642370ab6.png">

Reply:
It makes sense, although it can be a source of confusion to many users and we can't blame them. I'm confused myself :)

Reply:
Maybe we can work around this by providing some warnings if we detect any ambiguity.

Reply:
haha, me as well

logging a warning on conflicts would be a reasonable solution imo - i'll work on that and see what it looks like